src/connection.rs

```diff
@@ -8,793 +8,841 @@
+                w.wake()
+            }
         }
-        match self.mode {
-            Mode::Client => id.is_server(),
-            Mode::Server => id.is_client()
+
+        match &result {
+            Err(ConnectionError::Io(_)) => result,
+            Err(ConnectionError::Closed) => Ok(None),
+            _ => {
+                self.socket.close().await.or(Ok::<_, ConnectionError>(()))?;
+                result
+            }
         }
     }
 
-    fn add_pending(&mut self, frame: RawFrame) -> Result<(), ConnectionError> {
-        if self.pending.len() >= self.config.max_pending_frames {
-            return Err(ConnectionError::TooManyPendingFrames)
+    /// Get the next inbound `Stream` and make progress along the way.
+    ///
+    /// This is called from `Connection::next_stream` instead of being a
+    /// public method itself in order to guarantee proper closing in
+    /// case of an error or at EOF.
+    async fn next(&mut self) -> Result<Option<Stream>> {
+        loop {
+            self.garbage_collect().await?;
+
+            // For each channel and the socket we create a future that gets
+            // the next item. We will poll each future and if any one of them
+            // yields an item, we return the tuple of poll results which are
+            // then all processed.
+            //
+            // For terminated sources we create non-finishing futures.
+            // This guarantees that if the remaining futures are pending
+            // we properly wait until woken up because we actually can make
+            // progress.
+            //
+            // While it should never happen that all futures are terminated
+            // we nevertheless count them and return early if they are.
+
+            let mut num_terminated = 0;
+
+            let mut next_inbound_frame =
+                if self.socket.is_terminated() {
+                    num_terminated += 1;
+                    Either::Left(future::pending())
+                } else {
+                    Either::Right(self.socket.try_next().err_into())
+                };
+
+            let mut next_stream_command =
+                if self.stream_receiver.is_terminated() {
+                    num_terminated += 1;
+                    Either::Left(future::pending())
+                } else {
+                    Either::Right(self.stream_receiver.next())
+                };
+
+            let mut next_control_command =
+                if self.control_receiver.is_terminated() {
+                    num_terminated += 1;
+                    Either::Left(future::pending())
+                } else {
+                    Either::Right(self.control_receiver.next())
+                };
+
+            if num_terminated == 3 {
+                log::error!("{}: all futures are terminated", self.id);
+                return Err(ConnectionError::Closed)
+            }
+
+            let next_item =
+                future::poll_fn(move |cx: &mut Context| {
```

Comment:
There's a `maybe_done` wrapper that could probably do the same.
Something like that:

```
let a = future::maybe_done(next_stream_command);
let b = future::maybe_done(next_control_command);
let c = future::maybe_done(next_inbound_frame);
future::select(future::select(a, b), c).await;
if let MaybeDone::Done(a) = a {
    ...
}
// same for b and c
```

However your code might actually be more readable, so I'm not sure.
In general, selecting over multiple futures is still something that hasn't really figured out in the Rust async ecosystem yet, I think.